core/contracts/Voting.sol

```diff
@@ -607,56 +534,41 @@ contract Voting is Testable, MultiRole, OracleInterface, VotingInterface, Encryp
         return keccak256(abi.encode(identifier, time));
     }
 
-    /**
-     * @notice Updates the round if necessary. After this method is run voteTiming.getLastUpdatedRoundId() and
-     * and voteTiming.computeCurrentRoundId(blockTime) should return the same value.
-     * @dev The method loops through all price requests for the last voting round and rolls them over to the next round
-     * if required.
-     */
-    function _updateRound(uint blockTime) private {
-        if (!voteTiming.shouldUpdateRoundId(blockTime)) {
-            return;
+    function _getOrCreateSnapshotId(uint roundId) private returns (uint) {
+        Round storage round = rounds[roundId];
+        if (round.snapshotId == 0) {
+            // There is no snapshot ID set, so create one.
+            round.snapshotId = votingToken.snapshot();
         }
 
-        // Only do the rollover if the next round has started.
-        uint lastActiveVotingRoundId = voteTiming.getLastUpdatedRoundId();
-        Round storage lastActiveVotingRound = rounds[lastActiveVotingRoundId];
-
-        uint nextVotingRoundId = voteTiming.computeCurrentRoundId(blockTime);
-
-        for (uint i = 0; i < lastActiveVotingRound.priceRequestIds.length; i++) {
-            bytes32 priceRequestId = lastActiveVotingRound.priceRequestIds[i];
-            PriceRequest storage priceRequest = priceRequests[priceRequestId];
-
-            // TODO: we should probably take this assert out before we move to production to keep the voting
-            // contract from locking in the case of a bug. This would be an assert, but asserts don't allow
-            // messages.
-            require(priceRequest.lastVotingRound == lastActiveVotingRoundId,
-                "Found price request that was incorrectly placed in a round");
-
-            VoteInstance storage voteInstance = priceRequest.voteInstances[lastActiveVotingRoundId];
-
-            (bool isResolved, int price) = voteInstance.resultComputation.getResolvedPrice(
-                _computeGat(lastActiveVotingRoundId));
-
-            if (!isResolved) {
-                // If the vote cannot be resolved, push the request into the current round.
-                rounds[nextVotingRoundId].priceRequestIds.push(priceRequestId);
+        return round.snapshotId;
+    }
 
-                // Set the price resolution round number to the provided round.
-                priceRequest.lastVotingRound = nextVotingRoundId;
+    function _resolvePriceRequest(PriceRequest storage priceRequest, VoteInstance storage voteInstance) private {
+        if (priceRequest.isResolved) {
```

Comment:
Gas optimization that we can do later: wdyt about just setting the index to UINT_MAX and checking for that instead of storing isResolved separately?

Reply:
Yeah that would work. Done.